cmreactor/cmreactor.sh

```diff
@@ -0,0 +1,479 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <groupId>cmreactor</groupId>
+    <artifactId>cmreactor</artifactId>
+    <version>0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    <modules>
+EOROOTXMLSTART
+while [[ "$#" -gt 0 ]]; do
+    module="$1"
+    shift
+cat >> "${pomFile}" <<EOROOTXMLMOD
+        <module>${module}</module>
+EOROOTXMLMOD
+done
+cat >> "${pomFile}" <<EOROOTXMLEND
+    </modules>
+</project>
+EOROOTXMLEND
+}
+
+readConfigValue() {
+    local config
+    local key
+    local defaultValue
+
+    if [[ $# -lt 2 ]]; then
+        echo "readConfigValue requires two args: config key [optional default]" >&2
+        exit 1
+    fi
+
+    config="$1"
+    key="$2"
+    defaultValue="$3"
+
+    if gitc config -f "${config}" "${key}" >/dev/null; then
+        gitc config -f "${config}" "${key}"
+    elif [[ "$defaultValue" != "" ]]; then
+        echo "$defaultValue"
+    fi
+}
+
+readGitConfigSyncElements() {
+    local config
+    config="$1"
+
+    if gitc config -f "${config}" --get-all "cmreactor.configsync" >/dev/null; then
+        gitc config -f "${config}" --get-all "cmreactor.configsync"
+    else
+        echo user.email
+        echo user.name
+    fi
+}
+
+readModulesFromConfig() {
+    local config
+    config="$1"
+
+    if gitc config -f "${config}" --get-all "cmreactor.modules" >/dev/null; then
+        gitc config -f "${config}" --get-all "cmreactor.modules"
+    fi
+}
+
+readModuleConfigValue() {
+    local config
+    local module
+    local key
+    local defaultValue
+
+    if [[ $# -lt 3 ]]; then
+        echo "readModuleConfigValue requires three args: config module key [optional default]" >&2
+        exit 1
+    fi
+
+    config="$1"
+    module="$2"
+    key="$3"
+    defaultValue="$4"
+
+    if gitc config -f "${config}" "cmreactor.module.${module}.${key}" >/dev/null; then
+        gitc config -f "${config}" "cmreactor.module.${module}.${key}"
+    elif [[ "$defaultValue" != "" ]]; then
+        echo "$defaultValue"
+    fi
+}
+
+handleZipModule() {
+    local config
+    local module
+    local tempBase
+    local url
+    local zipFile
+    local extractedDir
+    local firstPomParentDir
+    local modulePath
+    local metaPath
+    local hash
+
+    config="$1"
+    module="$2"
+    tempBase="$3"
+
+    modulePath="$(pathInProject "${module}")"
+    url="$(readModuleConfigValue "${configFile}" "${module}" "url")"
+    zipFile="${tempBase}/${module}.zip"
+
+    if [[ "${url}" == http* ]]; then
+        curl -f -s -L -o "${zipFile}" "${url}"
+    elif [[ -f "${url}" ]]; then
+        cp "${url}" "${zipFile}"
+    else
+        return 1
+    fi
+    hash="$(git hash-object "${zipFile}")"
+    extractedDir="${tempBase}/${module}"
+    [[ -d "${extractedDir}" ]] && rm -rf "${extractedDir}"
```

Comment:
Is there any possibility at all that this could delete something important (without warning), given bad script arguments, duplicate module names, or anything else?

Reply:
I should probably filter the module names with a regular expression.

Reply:
Could "tempBase" be set to "" or "." and cause issues?

Reply:
nope. that's fully internal to the script, based on a constant path relative to the source root directory.

Reply:
true, thanks